qualtran/bloqs/arithmetic/Error propagation due to fixed point arithmetic.ipynb

```
@@ -0,0 +1,203 @@
+    "\n",
+    "$$\n",
+    "    |\\kappa\\rangle_{A} |\\lambda\\rangle_{B} |0\\rangle_{\\text{out}} \\rightarrow |\\kappa\\rangle_{A} |\\lambda\\rangle_{B} |\\widetilde{\\gamma}\\rangle_{\\text{out}}\n",
+    "$$\n",
+    "\n",
+    "\n",
+    "We analyze the error due to fixed point arithmetic for algorithm described in Appendix D5 of https://arxiv.org/abs/2007.07391"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "id": "b89122eb-d6b0-48f6-a3f4-4d3a0824ba7b",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Success! d=13, eps=0.001\n",
+      "Success! d=17, eps=0.0001\n",
+      "Success! d=21, eps=1e-05\n",
+      "Success! d=24, eps=1e-06\n",
+      "Success! d=28, eps=1e-07\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Multiplying two real numbers\n",
+    "def get_bitsize_for_fxp_mul(eps: float):\n",
+    "    return int(np.ceil(1 + np.log2(1/eps)) + np.log2(1 + np.log2(1/eps)))\n",
+    "\n",
+    "def mul_via_repeated_add(a: float, b: float, d: int):\n",
+    "    a_fxp, b_fxp = fxp(a, d=d), fxp(b, d=d)\n",
+    "    res = fxp(0, d=d)\n",
+    "    for i, b_bin in enumerate(b_fxp.bin()[:-1]):\n",
+    "        a_fxp >>= 1\n",
+    "        if int(b_bin):\n",
+    "            res += a_fxp\n",
+    "    return res\n",
+    "\n",
+    "def test_multiplication_for_eps(eps: float, d: int):\n",
+    "    rng = np.random.default_rng(int(eps * 1e9))\n",
+    "    try:\n",
+    "        for _ in range(100):\n",
+    "            a, b = rng.random(2)\n",
+    "            res = mul_via_repeated_add(a, b, d)\n",
+    "            assert_allclose(res, a * b, eps)\n",
+    "        print(f'Success! {d=}, {eps=}')\n",
+    "    except:\n",
```

bare except. Is this the best pattern here? asserting and catching?

---

changed to catch only `AssertionError` instead of bare except